service/notification/internal/app/event_handler.go

```diff
@@ -0,0 +1,166 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/broker"
+	broker_msgs "github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/broker/gen"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/config"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/entities"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/service"
+	"google.golang.org/protobuf/proto"
+)
+
+type (
+	Consumer interface {
+		ConsumeMessage(handler func(msg broker.ConsumedMessage)) error
+	}
+
+	NotificationService interface {
+		SendSubscriptionDetails(ctx context.Context, notification service.Notification) error
+		SendExchangeRates(ctx context.Context, notification service.Notification) error
+	}
+
+	DispatchStore interface {
+		AddSubscription(ctx context.Context, sub entities.Subscription) error
+	}
+
+	Scheduler interface {
+		AddSubscription(entities.Subscription)
+	}
+
+	eventHandler struct {
+		broker        Consumer
+		scheduler     Scheduler
+		logger        config.Logger
+		service       NotificationService
+		dispatchStore DispatchStore
+	}
+)
+
+const (
+	SubscriptionCreatedEvent string = "events.subscription.created"
+	SendDispatchCommand      string = "commands.send.dispatch"
+
+	TimeoutOfProcessing time.Duration = 2 * time.Second
+	RedeliveryDelay     time.Duration = 1 * time.Minute
+)
+
+func NewEventHandler(
+	broker Consumer,
+	dispatchScheduler Scheduler,
+	service NotificationService,
+	logger config.Logger,
+	dispatchStore DispatchStore,
+) *eventHandler {
+
+	return &eventHandler{
+		logger:    logger,
+		scheduler: dispatchScheduler,
+		broker:    broker,
+		service:   service,
+	}
+}
+
+func (h *eventHandler) handleSubscriptionCreatedEvent(msg broker.ConsumedMessage) error {
+	var parsedMsg broker_msgs.SubscriptionMessage
+	if err := proto.Unmarshal(msg.Data(), &parsedMsg); err != nil {
+		return fmt.Errorf("failed to unmarshal message from %s: %w", SubscriptionCreatedEvent, err)
+	}
+
+	sub := entities.Subscription{
+		DispatchID:  parsedMsg.Payload.DispatchID,
+		BaseCcy:     parsedMsg.Payload.BaseCcy,
+		TargetCcies: parsedMsg.Payload.TargetCcies,
+		Email:       parsedMsg.Payload.Email,
+		SendAt:      parsedMsg.Payload.SendAt.AsTime(),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), TimeoutOfProcessing)
+	defer cancel()
+
+	h.scheduler.AddSubscription(sub)
+	if err := h.dispatchStore.AddSubscription(ctx, sub); err != nil {
+		return fmt.Errorf("failed to save subscription: %w", err)
+	}
+
+	if err := h.service.SendSubscriptionDetails(ctx, service.Notification{
```

Comment:
A little bit confusing. For what does this responsible for, if you are sending mail not on subscription as i remember?

Reply:
You mean I should make naming of method more clear so it could be easy to understand context of subscription?